s4e-backend/src/main/java/pl/cyfronet/s4e/service/InvitationService.java

```diff
@@ -0,0 +1,117 @@
+                email,
+                institutionSlug,
+                Invitation.class
+        )
+                .orElse(null);
+        if (invitation != null) {
+            throw new InvitationCreationException("Invitation already exists in DB");
+        }
+
+        val token = UUID.randomUUID().toString();
+        invitation = Invitation.builder()
+            .email(email)
+            .token(token)
+                .status(InvitationStatus.WAITING)
+            .institution(institution)
+            .build();
+        invitationRepository.save(invitation);
+
+        return token;
+    }
+
+    @Transactional(rollbackFor = NotFoundException.class)
+    public void confirmInvitation(String token) {
+        val invitation = invitationRepository.findByToken(token, Invitation.class)
+                .orElseThrow(() -> new NotFoundException("Invitation with your token doesn't exist"));
+        val institution = invitation.getInstitution();
+
+        val user = appUserRepository.findByEmail(AppUserDetailsSupplier.get().getEmail())
+                .orElseThrow(() -> new NotFoundException("You're not login or your user don't exists"));
+        addMemberToDefaultGroup(user, institution);
+    }
+
+    @Transactional(rollbackFor = NotFoundException.class)
+    public void rejectInvitation(String token) {
+        val invitation = invitationRepository.findByToken(token, Invitation.class)
+                .orElseThrow(() -> new NotFoundException("Invitation with your token doesn't exist"));
+        invitation.setStatus(InvitationStatus.REJECTED);
+    }
+
+    public <T> Set<T> getAllBy(String institutionSlug, Class<T> projection) {
+        return invitationRepository.findAllByInstitution(institutionSlug, projection);
+    }
+
+    public <T> T findInstitutionBy(String token, Class<T> projection) {
+        val invitation = invitationRepository.findByToken(token, Invitation.class)
+                .orElseThrow(() -> new NotFoundException("Invitation couldn't be found"));
+        val institution = institutionRepository.findById(invitation.getId(), projection)
+                .orElseThrow(() -> new NotFoundException("Institution couldn't be found"));
+        return institution;
+    }
+
+    public <T> T findByToken(String token, Class<T> projection) {
+        return invitationRepository.findByToken(token, projection)
+                .orElseThrow(() -> new NotFoundException("Invitation with your token doesn't exists"));
+    }
+
+    public <T> Optional<T> findByEmailAndInstitutionSlug(String email, String invitationSlug, Class<T> projection) {
+        return invitationRepository.findByEmailAndInstitution(email, invitationSlug, projection);
+    }
+
+    public void remove(String token) {
+        val invitation = invitationRepository.findByToken(token, Invitation.class)
+                .orElseThrow(() -> new NotFoundException("Invitation with your token doesn't exists"));
+        invitationRepository.delete(invitation);
+    }
```

Comment:
This method should be `@Transactional`, as you remove an entityManager managed entity here.